core/pta/rtc.c

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2018, Linaro Limited
+ * Copyright (c) 2021, EPAM Systems. All rights reserved.
+ *
+ * Based on plat-synquacer/rtc_pta.c
+ *
+ */
+
+#include <drivers/rtc.h>
+#include <kernel/pseudo_ta.h>
+#include <rtc_pta_client.h>
+#include <tee_api_defines.h>
+
+#define PTA_NAME "rtc.pta"
+
+static TEE_Result rtc_pta_get_time(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct optee_rtc_time *time = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		DMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	time = params[0].memref.buffer;
+	if (!time || params[0].memref.size < sizeof(*time))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	rtc_get_time(time);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result rtc_pta_set_time(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct optee_rtc_time *time = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		DMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	time = params[0].memref.buffer;
+	if (!time)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	rtc_set_time(time);
```

Comment:
What happens if the info in `time` doesn't make sense?

Reply:
It actually depends on the driver. The AT91 RTC will actually return an error while programming it. Maybe I could add a sanity check to at least avoid total non sense but it will probably be a bit difficult (leap year, february duration and so on)